add brp helper scripts from mandriva's spec-helper

proyvind opened the pull request:
I've added most of the helper scripts, there's a few remaining and some I'll have to consider whether to kill off in stead before done.

Anyways, I'm making a preliminary pull request as a request for comments.

Notice how all these scripts can either be disabled by setting an environment variable or defining a rpm variable in contrast to the existing brp-* scripts.

As this has been an absolute necessity to disable some of these for some packages in cooker, I really think these should be kept, also equivalent disablers added for the existing brp- scripts invoked as well..

ignatenkobrain commented on platform.in:
```diff
@@ -70,6 +70,15 @@
+    %{?!dont_remove_info_dir:		[ -n "$DONT_REMOVE_INFO_DIR" ]	|| %{_rpmconfigdir}/brp-remove-info-dir} \
+    %{?!dont_relink:			[ -n "$DONT_RELINK" ]        	|| %{_rpmconfigdir}/brp-relink-symlinks} \
+    %{?!dont_remove_libtool_files:	[ -n "$DONT_REMOVE_LIBTOOL_FILES" ] || %{_rpmconfigdir}/brp-remove-libtool-files} \
+    %{?!dont_fix_eol:			[ -n "$DONT_FIX_EOL" ]		|| %{_rpmconfigdir}/brp-fix-eol} \
```
love this one ;) was fixingsuch things all the time manually ;)

ignatenkobrain commented on scripts/brp-remove-info-dir:
```diff
@@ -0,0 +1,15 @@
+
+dir="$RPM_BUILD_ROOT/$INFODIR"
+
+if [ -f $dir -a ! -f "$RPM_BUILD_ROOT/sbin/install-info" -a ! -f "$RPM_BUILD_ROOT/usr/sbin/install-info" ]; then
```
not sure what/why this one shoud do...

proyvind replied:
Some 'make install' rules uses 'install-info' to install info pages into install destination.
When installing info pages using 'install-info', it'll also register the info page in /usr/share/info/dir, for which in a clean buildroot doesn't exist and a new /usr/share/info/dir file with the info page entry gets created.

So this ensures that no /usr/share/info/dir ends up in package, conflicting with existing system /usr/share/info/dir

Conan-Kudo replied:
That's useful, since we currently make packagers purge that directory manually, which is annoying.

ignatenkobrain commented on scripts/brp-remove-libtool-files:
```diff
@@ -0,0 +1,14 @@
+	exit 0
+fi
+
+find "$RPM_BUILD_ROOT" \( -type f -o -type l \) -name \*.la  -print0 |
```
nice one ;)

Conan-Kudo replied:
Yay!

ignatenkobrain commented on scripts/brp-remove-rpath:
```diff
@@ -0,0 +1,49 @@
+fi
+
+# make sure we have chrpath in our search path, otherwise wuzz out
+type chrpath || exit 0
```
wondering how people are using this brp.... it's nice one, but you don't know in advance if built thing will contain rpaths or not..

proyvind replied:
It'll only check whether binaries contains a RPATH or RUNPATH tag after finishing %install, so no need to know in advance.

As it pulls in a dependency on chrpath which might not be desired/available to have installed by default in build environment, it'll just wuzz out without checking nor fixing anything.

Conan-Kudo replied:
I'm not quite sure if this one is a good idea, as there are plenty of cases where RPATH/RUNPATH makes sense to preserve (such as when they are private/plugin libs in `/usr/lib64/<progdir>`).

proyvind replied:
You misunderstand the script's purpose.
It doesn't unconditionally remove all rpaths, but rather rpaths for standard search directories such as ie. /usr/lib64 as rpath, which first of all doesn't make sense, secondly it takes presence over $LD_LIBRARY_PATH, breaking the possibility for overriding default search path for binaries with such rpaths.

ldv-alt commented on scripts/brp-remove-libtool-files:
```diff
@@ -0,0 +1,14 @@
+find "$RPM_BUILD_ROOT" \( -type f -o -type l \) -name \*.la  -print0 |
+xargs  --no-run-if-empty -0 file -N -L |
+grep "libtool library file" |
+sed -e 's#: libtool library file.*##g' |
```
grep followed by sed? Really? Is it an accepted shell scripting style nowadays?

proyvind replied:
well, to my defense, shell scripting receives little focus and not exactly the language of choice I spend the most time when doing development. Also I've never really grown comfortable with regexp and I truly *hate* perl,  which I rewrote several of these scripts in shell from.. :p 

But at least it does what it's supposed to do and works well at it, so that's good enough for me as a start, but I'd love to see others applying a more sexy final touch to improve the scripts. :o)

proyvind replied:
For the brp-check-elf-files script, it's hard to contest that more work has been put into yours than pixel (I'm guessing) put into writing the original perl script, which I've simply rewritten in bash.

Yet for the brp- scripts, I'm a firm believer in KISS, where I really think the brp* scripts should be self-contained and for single purpose only with as little complexity as possible.

Conan-Kudo replied:
For what it's worth, the ALT Linux scripts are pretty hard to figure out, and the original Mandriva ones are not much better. These simpler ones are much easier to mentally parse and digest.

Conan-Kudo commented on platform.in:
```diff
@@ -66,10 +66,21 @@
+    %{?!dont_fix_eol:			[ -n "$DONT_FIX_EOL" ]		|| %{_rpmconfigdir}/brp-fix-eol} \
+    %{?!dont_remove_rpath:		[ -n "$DONT_REMOVE_RPATH" ]	|| %{_rpmconfigdir}/brp-remove-rpath} \
+    %{?!dont_fix_file_permissions: 	[ -n "$DONT_FIX_FILE_PERMISSIONS" ] || %{_rpmconfigdir}/brp-fix-file-permissions} \
+    %{?!dont_fix_xdg:			[ -n "$DONT_FIX_XDG" ]       	|| %{_rpmconfigdir}/brp-fix-xdg} \
```
This should be renamed to something more obvious. Maybe `fix_desktop_files`. The name "XDG" isn't very helpful...

proyvind commented:
For the rpath & pkgconfig, they've been well tested and various breakages in *cooker has all been fixed with all issues encountered having been properly fixed without any issues being reported for some years now. :)

And considering that you suggest that they should rather be explicily enabled rather than by default, I don't see any reason for dropping, it's entirely optional to enable.

Conan-Kudo commented:
@proyvind If you are changing them to be disabled by default, then I don't particularly care if they are there, I suppose.

proyvind commented:
could you describe a scenario where any of these brp-scripts wouldn't be desired to have enabled by default?

pmatilai commented:
I've nothing against adding more brp-scripts as such, I haven't looked at this in any detail but there seems to be useful stuff in there. 

Enabling such a big pile of new stuff by default is a wholly different question, but lets not get hung up on that. The other complaint wrt enablement is that "dont" as a disabler doesn't sound right, "brp" stands for build root policy and enable/disable seem more fitting terminology for that. On a related note, whatever the means of enabling and disabling these policies ends up being, it needs to cover the pre-existing policies too, but that's beyond the scope of this PR. Neither of these issues prevents getting the scripts themselves in.

I do have some issues with the script names, they're fairly hodge-podgey and not very descriptive in places. Not that the ones currently upstream truly follow any defined pattern but lets not make that situation worse.

On a related note, looking at this many brp-scripts at once makes it painfully obvious how stupid the whole mechanism is. Every script repeats the same work over and over again, so we end up checking for buildroot and running find after find after find. I'd think all this could be coupled with the file classifier: eg the file classifier already knows if there are ELF files in the buildroot, and we could use that info to only run ELF-related brp-scripts when actually needed. Etc.

soig commented:
On 24 February 2017 at 11:32, Panu Matilainen <notifications@github.com>
wrote:

> I've nothing against adding more brp-scripts as such, I haven't looked at
> this in any detail but there seems to be useful stuff in there.
>
> Enabling such a big pile of new stuff by default is a wholly different
> question, but lets not get hung up on that. The other complaint wrt
> enablement is that "dont" as a disabler doesn't sound right, "brp" stands
> for build root policy and enable/disable seem more fitting terminology for
> that. On a related note, whatever the means of enabling and disabling these
> policies ends up being, it needs to cover the pre-existing policies too,
> but that's beyond the scope of this PR. Neither of these issues prevents
> getting the scripts themselves in.
>
> I do have some issues with the script names, they're fairly hodge-podgey
> and not very descriptive in places. Not that the ones currently upstream
> truly follow any defined pattern but lets not make that situation worse.
>
> On a related note, looking at this many brp-scripts at once makes it
> painfully obvious how stupid the whole mechanism is. Every script repeats
> the same work over and over again, so we end up checking for buildroot and
> running find after find after find. I'd think all this could be coupled
> with the file classifier: eg the file classifier already knows if there are
> ELF files in the buildroot, and we could use that info to only run
> ELF-related brp-scripts when actually needed. Etc.
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/rpm-software-management/rpm/pull/122#issuecomment-282259094>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AC5qkeGIrHRNsV3jrs_iWkTCePMoKRo4ks5rfrGngaJpZM4LhDpA>
> .
>

​Which reminds me (on a similar note) that we should really make​ the "new
dependency generator" faster by making it able to call the commands listed
in .attr file once for all files (could be an extra flag but I'm pretty
sure all default scripts are OK b/c of history (compatibility with old
dependency generator))

proyvind commented:
@soig you're making a good point, if one wanna look at truly stupid mechanisms, one should look at how the internal dependency generator is doing the same work over and over and over.. It's now become what takes up the most time of build for certain packages..

In contrast, running 'find' over and over within a buildroot with caches enabled and all, the amount of work being repeated is neglible, just as the time it takes to run the scripts.
The scripts are written in a similar fashion as the existing ones in place, where limiting the amount of times 'find' being run really hasn't been of much relevance, these are simple shell scripts.
Refactoring of the scripts really aren't all that worthwhile spending time on, especially not compared to the gain of the functionality provided by.
For health reasons, I'm only sporadically active, so it's certainly not something worthwhile spending my time on at least.

Same for script names, does it matter all that much?
The names comes from the mandriva spec-helper package, where they originate from.
If not clear enough, it sounds more sensible to just document it's behaviour in the scripts themself anyways..

If the the scripts and functionality of the scripts are sensible enough, I recommend just applying whatever cosmetics on top of my pull request as-is, I'm rather absent these days.

proyvind commented:
IMO the reasonable concerns has been addressed now and this PR should be ready to be merged.

proyvind commented:
considering that the brp scripts is amongst the most important PR improving packaging life in ages, it starts to reflect poorly of not just not giving it priority, but actually just simply ignoring it for several months now.

last brp script with purpose not reflected by file name has been fixed in new commits pushed, please merge these ASAP or make at least a good argument against doing so, this is starting to get unreasonable..

Conan-Kudo commented on scripts/brp-lib-symlinks:
```diff
@@ -0,0 +1,25 @@
+#
+# Copyright (C) 2000 by Frederic Lepied <flepied@mandrakesoft.com>
+# Redistribution of this file is permitted under the terms of the GNU
+# Public License (GPL).
```
What does this mean? What version of the license?

Conan-Kudo commented on scripts/brp-relink-symlinks:
```diff
@@ -0,0 +1,42 @@
+#
+# Copyright (C) 2012 by Per Øyvind Karlsen <proyvind@mandriva.org>
+# Redistribution of this file is permitted under the terms of the GNU
+# Public License (GPL).
```
Same as earlier.

Conan-Kudo commented on scripts/brp-remove-info-dir:
```diff
@@ -0,0 +1,25 @@
+#
+# Copyright (C) 2011 by Per Øyvind <peroyvind@mandrake.org>
+# Redistribution of this file is permitted under the terms of the GNU
+# Public License (GPL).
```
Same as earlier.

Conan-Kudo commented on scripts/brp-remove-libtool-files:
```diff
@@ -0,0 +1,24 @@
+#
+# Copyright (C) 2011 by Per Øyvind <peroyvind@mandrake.org>
+# Redistribution of this file is permitted under the terms of the GNU
+# Public License (GPL).
```
Same as earlier.

Conan-Kudo commented on scripts/brp-remove-rpath:
```diff
@@ -0,0 +1,58 @@
+#
+# Copyright (C) 2013 by Per Øyvind <peroyvind@mandrake.org>
+# Redistribution of this file is permitted under the terms of the GNU
+# Public License (GPL).
```
Same as earlier.

proyvind commented:
GitHub feature request: Ignore PRs
...

ignatenkobrain commented on scripts/brp-fix-desktop-files:
```diff
@@ -0,0 +1,17 @@
+#!/bin/sh
+# Fix .desktop files to be compliant with XDG specification
```
please, don't ;)

proyvind replied:
Why? .desktop files issuee corrected by this script is quite widespread and goes often unnoticed.

Trivial to automatically fix with script, tedious to manually pursue fixing...

proyvind commented:
For the package exceptions requiring the libtool .la files, you have a disabler macro.
That's the whole point of the disabler macros in general, while enabling the scripts are desired in most cases, thus enabled by default, there always might be exceptions where not desired and for those you can simply use the disablers.

And disabler macros are added for scripts for consistency (regardless of perceived usefulness of disabling any particular behaviour of individual scripts).

proyvind commented:
No idea what you mean by prefering users explicitly doing the right thing..
Trying to interpret it, it seems to contradict the purpose of these scripts in general, avoiding users having to explicitly take care of things that can be safely automized and handled automatically outside of spec files.
Requiring packagers to explictly do the right thing is not a good argument as you really wanna avoid the chance of human errors (which are the most common error of all) as much as possible.

brp-relink-symlinks was written long time ago and has been in production use since without any issues.

For "too magic", I'd rather think of it as doing it in the most reliable/correct fashion.

It would be far more constructive with comments about the scripts behaviour, purpose and how implemented rather than raising concerns based on false assumptions about the condition of.

proyvind commented:
brp-clean-files is convenient as it'll simply make sure that files which surely has no place in builldroot are removed and accidental packaging of is prevented.

The script is extremely simple and the most trivial one in this PR, so I fail to see your concern about it might break something.

proyvind commented:
considering how long this PR has stayed open without any more feedback from upstream, I'm wondering whether if fixing them would be a waste of time or not..?

Conan-Kudo commented:
@proyvind We now have a consistent form of how to enable/disable brp scripts, so reworking this to leverage that would be appreciated.

n3npq commented:
pmatlai has added a mechanism to enable/disable specific brp scripts within opt-in or opt-out packager control in a *.spec build recipe.

Further automation runs into the usual chicken-egg problem, that the files to be included in a package are not known until the %buildroot staging area has been constructed, which also require changing the way that brp scripts are written to handle a fine grained triggered execution control.

Consistent usage is a matter of distro consensus and usage, not otherwise.

proyvind commented:
by fixing conflicts, I implicitly meant adapting the brp- scripts to panu's latest changes.

I was wondering if it's a waste of time or not, given that no further discussion took place since having addressed all concerns raised, it's been over a year ago, making me questioning whether it'll still would continue remain ignored?

pmatilai commented:
Ok, this really does need a status update.

We're not going to merge this as-is in any case, so don't waste your time updating the PR for the new brp-disabling style or otherwise. It's impossible to properly review this many scripts in one PR, and  distros would be really unhappy with such a huge pile of possibly build-breaking changes at once. Distros need to manage such things at their own pace.

What we *will* do is we'll look at what the various distros are doing in this regard and merge the best of the most common things. For example automated ldconfig is pretty much everywhere but everybody's been reinventing the wheel. That just doesn't make any sense.

Second, we think there should be a kind of contrib repo for the rest  (ie the less common bits) with looser rules of what can be added, from which distros can pick and choose as they see fit but there would then at least be a common place for the stuff. Same goes for dependency generators BTW.

ffesti commented:
Added the brp scripts to the rpm-extras repo. Further discussion and refinement can happen there. Closing the PR here.
